### PR TITLE
Graduate Indexed Job to beta

### PR DESCRIPTION
**Pull request by alculquicondor**
Ref #2214

/sig apps

### CODE REVIEW

File: keps/sig-apps/2214-indexed-job/README.md
Comment (soltysh) [inline keps/sig-apps/2214-indexed-job/README.md:429]: Any particular reasons why these metrics will be for Indexed Jobs only? IMO it seems reasonable to expose them for all jobs.

File: keps/sig-apps/2214-indexed-job/README.md
Comment (soltysh) [inline keps/sig-apps/2214-indexed-job/README.md:463]: I'd like to have clear information what will be counted as an error towards this SLO? For example, I wouldn't include update conflicts as errors. Maybe it's reasonable to either explain what is an error for this metric?

File: keps/sig-apps/2214-indexed-job/README.md
Comment (soltysh) [inline keps/sig-apps/2214-indexed-job/README.md:466]: Do you have any metrics already? Few tens of pods sync shouldn't be longer than 10s.

File: keps/sig-apps/2214-indexed-job/README.md
Comment (soltysh) [inline keps/sig-apps/2214-indexed-job/README.md:467]: That 1h is definitely too long for a single sync.

**Review by soltysh — COMMENTED** [review]
I left a few questions.

File: keps/sig-apps/2214-indexed-job/README.md
Comment (alculquicondor) [reply keps/sig-apps/2214-indexed-job/README.md:429]: I don't plan to add them just for Indexed Jobs, but there will be a label for the mode.

File: keps/sig-apps/2214-indexed-job/README.md
Comment (alculquicondor) [reply keps/sig-apps/2214-indexed-job/README.md:466]: This is based off the pod creation SLOs, which are at 1s per pod https://github.com/kubernetes/community/blob/master/sig-scalability/slos/slos.md#steady-state-slisslos

In general, these number are very conservative, as we don't know in which kind of machine the control plane will be running.

File: keps/sig-apps/2214-indexed-job/README.md
Comment (alculquicondor) [reply keps/sig-apps/2214-indexed-job/README.md:463]: I was just thinking of whether the `syncJob` method returns error. But we can be more restrictive. We can discuss in SIG meeting.

File: keps/sig-apps/2214-indexed-job/README.md
Comment (alculquicondor) [reply keps/sig-apps/2214-indexed-job/README.md:467]: not if you consider https://github.com/kubernetes/community/blob/master/sig-scalability/slos/slos.md#steady-state-slisslos

File: keps/sig-apps/2214-indexed-job/README.md
Comment (alculquicondor) [reply keps/sig-apps/2214-indexed-job/README.md:463]: We didn't get a chance to discuss this in the meeting.

I'm thinking that this metric is a tool for cluster operators to quickly see if things as working as expected in dashboards and easily set alerts.

Let's say there is a rogue controller that is updating job pods, causing conflicts all the time. If we silence Conflict errors, an operator might think that the cluster is operating normally, as no alerts are triggered.

We should not think of this metric as "there is a bug in the controller" (although it can also be used for that), but about making sure that the controller is progressing and there are no agents degrading its operations.

File: keps/sig-apps/2214-indexed-job/README.md
Comment (alculquicondor) [reply keps/sig-apps/2214-indexed-job/README.md]: Updated to reflect that users can use a label to compare Indexed to NonIndexed

File: keps/sig-apps/2214-indexed-job/README.md
Comment (alculquicondor) [reply keps/sig-apps/2214-indexed-job/README.md:471]: From @wojtek-t 

> What exactly "job_sync_duration_seconds" is reporting?

> Is it reporting the duration of a single "sync in a controller"? Or processing a job?
If the former, it should be fairly fast right? [I'm assuming we're creating pods themselves asynchronously?]

It is reporting a single sync in a controller. We create/delete pods in parallel, but we wait for all of them to succeed. What would be the recommendation? Can we simply assume that all of them would finish within 1s?

Comment (wojtek-t) [reply keps/sig-apps/2214-indexed-job/README.md:475]: > It is reporting a single sync in a controller. We create/delete pods in parallel, but we wait for all of them to succeed.

Thanks for explanation.
We can't really assume they will finish fast then, as we still have QPS limits on the client-side.

The problem I see here, is that if a single sync can take that long, then one very large job (or N very large jobs where N=number of workers of the controller) can actually starve everything else for a really long time. That's far from perfect.

What we should do is to ensure that in a single sync we create no more than N pods (say 500 - I think 500 is used by ReplicaSet controller). And that gives you much better predictability (you can say that a single sync always has to finish within M seconds).
That has a drawback that it adds some latency if you indeed would be able to create e.g. 10k pods at once, but I think the impact is relatively small and you solve the starvation problem.

File: keps/sig-apps/2214-indexed-job/README.md
Comment (wojtek-t) [reply keps/sig-apps/2214-indexed-job/README.md:476]: I agree that 1h is waay to much (and I don't really see the relevance with the link you pasted Aldo).

Also see - my comment above (it would help addressing this one too).

File: keps/sig-apps/2214-indexed-job/kep.yaml
Comment (wojtek-t) [inline keps/sig-apps/2214-indexed-job/kep.yaml:23]: Update stage above

**Review by wojtek-t — COMMENTED** [review]
I'm mostly fine with is modulo the SLO-related comment I added.

File: keps/sig-apps/2214-indexed-job/README.md
Comment (alculquicondor) [reply keps/sig-apps/2214-indexed-job/README.md:475]: I'm fine with adding a limit to the number of created pods. Should we do it for deletions too (happens when users reduces parallelism or the job reached failure limits)?

Then for 500 pods, assuming 50 calls per second, we are talking about 10s for pods + 1s for Job update. Sounds reasonable? This will increase a bit with job tracking with finalizers, but we can discuss that in its beta graduation.

File: keps/sig-apps/2214-indexed-job/kep.yaml
Comment (alculquicondor) [reply keps/sig-apps/2214-indexed-job/kep.yaml:23]: Will do in next update :)

File: keps/sig-apps/2214-indexed-job/README.md
Comment (wojtek-t) [reply keps/sig-apps/2214-indexed-job/README.md:475]: Yes - we should do for deletions too.

O(10s) SLO sounds waaay better than 1m or 1h.

> This will increase a bit with job tracking with finalizers, but we can discuss that in its beta graduation.

It adds one additional call per 500 calls, right? So I would say it's negligible.

Comment (alculquicondor) [reply keps/sig-apps/2214-indexed-job/README.md:475]: Right, big jobs are barely affected. And indexed jobs can actually skip some calls :)

In any case, I'll add SLOs for parallelism=1, 50 and 500

Comment (wojtek-t) [reply keps/sig-apps/2214-indexed-job/README.md:475]: > Right, big jobs are barely affected. And indexed jobs can actually skip some calls :)

Smaller jobs are not affected at all, because they will never not create a 500-item batch.
So only large jobs may be affected.

Comment (wojtek-t) [reply keps/sig-apps/2214-indexed-job/README.md:475]: > In any case, I'll add SLOs for parallelism=1, 50 and 500

I would really like to have one SLO (without any steps).

Comment (alculquicondor) [reply keps/sig-apps/2214-indexed-job/README.md:475]: > Smaller jobs are not affected at all

we still have to remove finalizers from the pods and job (but we are going off topic for the purpose of this PR).

> I would really like to have one SLO (without any steps).

will do, assuming 500 operations limit.

Comment (alculquicondor) [reply keps/sig-apps/2214-indexed-job/README.md]: updated

File: keps/sig-apps/2214-indexed-job/README.md
Comment (wojtek-t) [inline keps/sig-apps/2214-indexed-job/README.md:473]: Let's not be that aggressive maybe. 15s?

File: keps/sig-apps/2214-indexed-job/README.md
Comment (wojtek-t) [inline keps/sig-apps/2214-indexed-job/README.md:474]: BTW - can you please also update the KEP to say that we need to divide those large sync into batches of 500 (or some other N if you have opinion on exact number).

**Review by wojtek-t — COMMENTED** [review]
Just some smaller comments.

Also waiting for @soltysh to approve first from the SIG perspective before approving PRR.

File: keps/sig-apps/2214-indexed-job/README.md
Comment (alculquicondor) [reply keps/sig-apps/2214-indexed-job/README.md:474]: it's right below :)

**Review by alculquicondor — COMMENTED** [review]
Updated SLO to 15s and squashed

File: keps/sig-apps/2214-indexed-job/README.md
Comment (wojtek-t) [inline keps/sig-apps/2214-indexed-job/README.md:478]: I think we should move it to the design section, rather than hide it here.

**Review by wojtek-t — COMMENTED** [review]
I'm generally fine with this modulo this minor comment - waiting for @soltysh approval first though.

File: keps/sig-apps/2214-indexed-job/README.md
Comment (alculquicondor) [reply keps/sig-apps/2214-indexed-job/README.md]: Moved to risks and mitigations

File: keps/sig-apps/2214-indexed-job/README.md
Comment (alculquicondor) [reply keps/sig-apps/2214-indexed-job/README.md:442]: I think mode is enough. @logicalhan is there any preference between shorter or more descriptive label names?

Comment (alculquicondor) [reply keps/sig-apps/2214-indexed-job/README.md:442]: ah, missed that. I homogenized to `mode` for now. We can always change it during code review if people feel strongly about the long name.

Comment (soltysh) [reply keps/sig-apps/2214-indexed-job/README.md:442]: My vote would be towards `completion_mode` since this is consistent with the API, but I won't block this KEP on it, it can be updated afterwards.

**Review by soltysh — APPROVED** [review]
/lgtm
/approve

**Comment by chrisnegus** [comment]
Hi @alculquicondor  ! I'm @chrisnegus , a Docs Shadow for release 1.22.
This enhancement is marked as Needs Docs for the 1.22 release.

Please follow the steps detailed in the documentation to [open a PR against dev-1.22 branch](https://kubernetes.io/docs/contribute/new-content/new-features/#open-a-placeholder-pr) in the kubernetes/website repo. This PR can be just a placeholder at this time and must be created before *Fri July 9, 11:59 PM PDT*.

Also, take a look at [Documenting for a release](https://kubernetes.io/docs/contribute/new-content/new-features/#for-developers-or-other-sig-members) to familiarize yourself with the docs requirement for the release.

Thank you!

**Pull request merged**
